crossbar/worker/proxy.py

```diff
@@ -0,0 +1,537 @@
+        self.session = session  # the authenticator (among others?) will call on this session
+
+
+class ProxyRouter(object):
+    """
+    Provide enough Router functionality so that crossbar's base
+    RouterSession class can call various methods on us..
+    """
+    log = txaio.make_logger()
+
+    def __init__(self, realm, worker, transport_config):
+        self._realm = ProxyRealm(realm, worker)
+        self._worker = worker
+        self._front_to_back = dict()  # session_id -> ProxySession backend
+        self._backend_transport_config = transport_config
+
+    def has_role(self, role):
+        """
+        we just forward any role at all onto the 'real' backend
+
+        XXX TODO: is this true? or should we config which roles are
+        okay? or is that up to 'the real backend' itself?
+        """
+        return True
+
+    def _make_backend_connection(self, realm, authid, authrole):
+        backend_connected_d = Deferred()
+
+        self.log.info(
+            "backend connection to: {config}",
+            config=self._backend_transport_config,
+        )
+        ep = create_connecting_endpoint_from_config(
+            self._backend_transport_config['endpoint'],
+            self._worker._cbdir,
+            self._worker._reactor,
+            self.log,
+        )
+
+        def create_session():
+            self.log.info("ProxyRouter connected, creating session", )
+            # backend_cfg = list(self._worker._backend_configs.values())[0]
+            s = Session(types.ComponentConfig(realm=realm))
+            s.add_authenticator(AuthProxy(
+                authrealm=realm,
+                authid=authid,
+                authrole=authrole,
+            ))
+
+            def _join(*args):
+                self.log.info("ProxyRouter joined to backend session", )
+                backend_connected_d.callback(s)
+
+            s.on('join', _join)
+            return s
+
+        # XXX FIXME TODO look at transport-type etc
+        factory = WampWebSocketClientFactory(create_session, url="ws://localhost:9000/ws")
+        conn_d = ep.connect(factory)
+        conn_d.addErrback(backend_connected_d.errback)
+        return backend_connected_d
+
+    def _session_joined(self, session, details):
+        self.log.info(
+            "ProxyRouter _session_joined {details}",
+            details=details,
+        )
+        frontend = session
+        # connect to the backend
+        authid = session._authid
+        authrole = session._authrole
+        realm = session._realm
+        d = self._make_backend_connection(realm, authid, authrole)
+
+        def got_backend(backend):
+            self._front_to_back[details['session']] = backend
+
+            def backend_connected(_, details):
+                assert backend._session_id is not None
+
+                # always just forward to frontend
+                backend.onMessage = frontend._transport.send
+                frontend.set_backend(backend)
```

Comment:
I'm not 100% sure this is correct behavior -- are there any message-kinds we **don't** want to forward (like the `do_not_forward` going the other way)...?